os/fs/smartfs/smartfs_smart.c

```diff
@@ -567,10 +577,45 @@ static ssize_t smartfs_read(FAR struct file *filep, char *buffer, size_t buflen)
 		}
 	}
 
-	/* Return the number of bytes we read */
-
+	final_output = (char *)malloc(buflen + 1);
```

Comment:
I think it is better to add a new module for ex: smartf_sfs.c which will all the implementation and only very few calls and minor changes will be done to smartfs_smart.c.. 
It may not be good to modify too much inside the existing smartfs_smart implementation.